xcpmd/src/xcpmd-dbus-server.c

```diff
@@ -476,64 +460,61 @@ gboolean xcpmd_battery_is_present(XcpmdObject *this, guint IN_bat_n, gboolean *O
 
 gboolean xcpmd_battery_state(XcpmdObject *this, guint IN_bat_n, guint *OUT_state, GError **error)
 {
-    /* 0: Unknown */
-    /* 1: Charging */
-    /* 2: Discharging */
-    /* 3: Empty */
-    /* 4: Fully charged */
-    /* 5: Pending charge */
-    /* 6: Pending discharge */
-
-    int juice_left;
-    int juice_when_full;
-    int percent;
-    unsigned int i;
-
     if (battery_slot_exists(IN_bat_n) != YES) {
         g_set_error(error, DBUS_GERROR, DBUS_GERROR_FAILED, "No such battery slot: %d", IN_bat_n);
         return FALSE;
     }
 
-    if (last_status[IN_bat_n].state & 0x1)
-        *OUT_state = 2;
-    else if (last_status[IN_bat_n].state & 0x2)
-        *OUT_state = 1;
-    else {
-        /* We're not charging nor discharging... */
-        juice_left = last_status[IN_bat_n].remaining_capacity;
-        juice_when_full = last_info[IN_bat_n].last_full_capacity;
-        percent = juice_left * 100 / juice_when_full;
-        /* Are we full or empty? */
-        if (percent > 90)
-            *OUT_state = 4;
-        else if (percent < 10)
-            *OUT_state = 3;
-        else {
-            /* Is anybody else (dis)charging? */
-            for (i = 0; i < num_battery_structs_allocd; ++i) {
-                if (i != IN_bat_n &&
-                    last_status[i].present == YES &&
-                    (last_status[i].state & 0x1 || last_status[i].state & 0x2)) {
-                    break;
-                }
-            }
-            if (i < num_battery_structs_allocd) {
-                /* Yes! */
-                /* If the other battery is charging, we're pending charge */
-                if (last_status[i].state & 0x2)
-                    *OUT_state = 5;
-                /* If the other battery is discharging, we're pending discharge */
-                else if (last_status[i].state & 0x1)
-                    *OUT_state = 6;
-            } else
-                /* We tried everything, the state is unknown... */
-                *OUT_state = 0;
-        }
```

Comment:
The function get_total_charge_rate() in battery.c needs to know the current charge state of each battery individually.